### PR TITLE
Containerizing OpenManage Ansible Modules

### PR DESCRIPTION
**Pull request by saishivarcr**
Added Dockerfile and README.md file with instructions on containerizing Dell EMC OpenManage Ansible Modules

### CODE REVIEW

**Comment by mariolenz** [comment]
Personally, I think it would be better if you install these modules with `ansible-galaxy collection install dellemc.openmanage` instead of curling the latest realease from GitHub and installing it more or less manually.

Anyway, I'm working in an enterprise environment (like most Dell Customers, I think) and I don't see our IT Security allowing us to run a container based on something from [DockerHub](https://hub.docker.com/_/python). I can see use cases to put this into a container, but I don't think this should be done in this repo.

**Review by rajeevarakkal — CHANGES_REQUESTED** [review]
Open Manage does holds maintaining rights of the Image customer Creates, remove this option LABEL

**Review by rajeevarakkal — COMMENTED** [review]
Any reasons why are not using clone the branch and run install command? Is git package not supported in slim version?

**Review by rajeevarakkal — COMMENTED** [review]
Do this need the control machine connected and has access to Docker hub during `run` as well?

File: docker/Dockerfile
Comment (rajeevarakkal) [inline docker/Dockerfile:1]: Do this need the control machine connected and has access to Docker hub during run as well?

File: docker/Dockerfile
Comment (rajeevarakkal) [inline docker/Dockerfile:10]: Any reasons why are not using clone the branch and used install command? Is git package not supported in slim version?

File: docker/Dockerfile
Comment (rajeevarakkal) [inline docker/Dockerfile:2]: Open Manage does holds maintaining rights of the Image customer Creates, remove this option LABEL

**Review by rajeevarakkal — CHANGES_REQUESTED** [review]

File: docker/README.md
Comment (rajeevarakkal) [inline docker/README.md:6]: Can you provide the short description of the each packages used such as base images and ansible packages used in the `Dockerfile`. ? Also the Disk size required.

**Review by rajeevarakkal — CHANGES_REQUESTED** [review]

File: docker/README.md
Comment (rajeevarakkal) [inline docker/README.md:9]: Can you add a disclaimer to remind the respective data center administrator to get the required approvals for consuming the docker images and other packages from open source repositories?

File: docker/README.md
Comment (rajeevarakkal) [inline docker/README.md:16]: Provide the overview on where administrator can find the OMAM playbook example and how to use it

**Review by rajeevarakkal — CHANGES_REQUESTED** [review]

**Comment by anupamaloke** [comment]
> Personally, I think it would be better if you install these modules with `ansible-galaxy collection install dellemc.openmanage` instead of curling the latest realease from GitHub and installing it more or less manually.
> 
> Anyway, I'm working in an enterprise environment (like most Dell Customers, I think) and I don't see our IT Security allowing us to run a container based on something from [DockerHub](https://hub.docker.com/_/python). I can see use cases to put this into a container, but I don't think this should be done in this repo.

@mariolenz, the intent behind the Dockerfile is to let uses have a flexibility to use containers for running for running the Ansible modules. And you are right that there is very very low chance of someone running this in production. This is more targeted towards development /pre-production environment where you want to try a change quick and fast. Of course, **ansible-collections** are still the **de facto** way.  
Just my two-cents :-)

File: docker/README.md
Comment (rajeevarakkal) [inline docker/README.md:26]: Do we need to access Docker Host Resources/artifact from docker host frequently? How do we planning to manage share access required for modules such export system configuration profiles . Will this share created already inside container?

File: docker/Dockerfile
Comment (mariolenz) [reply docker/Dockerfile:10]: Any reason at all to fetch the code? I think the workflow will look like this: 1) clone this repo 2) build the image.

You see, everything's already available locally. I'm no container expert, but can't you build the modules from the cloned repository and copy them to the docker image?

**Comment by mariolenz** [comment]
> the intent behind the Dockerfile is to let uses have a flexibility to use containers for running for running the Ansible modules. And you are right that there is very very low chance of someone running this in production. This is more targeted towards development /pre-production environment where you want to try a change quick and fast. Of course, **ansible-collections** are still the **de facto** way.

I think I didn't make myself clear, I didn't mean to use `ansible-galaxy collection install dellemc.openmanage` _instead_ of building a container, I meant to use ansible-galaxy in the dockerfile to install the modules. But if this is more targeted towards development / pre-production, I can see this wouldn't make sense. In that case, you want the latest code from devel instead of the latest (stable) code released on galaxy.

File: docker/Dockerfile
Comment (saishivarcr) [reply docker/Dockerfile:8]: ansible 2.10 is not available on PyPI yet, once it is published we can test it out and do the required changes as you have suggested.

File: docker/Dockerfile
Comment (saishivarcr) [reply docker/Dockerfile:10]: @rajeevarakkal Yes git is not installed on 'slim' version

@mariolenz These were the thoughts running on my mind while designing it:
* Customer need not clone the entire repository in order to use it, he can just copy this Dockerfile and build it on hist own. Also planning to push pre-built image to docker hub which also eliminates need to clone the repository from github.
* In future we are planning to integrate with Dockerhub, so that we can enable automated build generation with appropriate tagging in place.

With these in mind, I thought of writing standalone Dockerfile which is capable of generating image without any other dependencies.

File: docker/Dockerfile
Comment (saishivarcr) [reply docker/Dockerfile:2]: @rajeevarakkal I am OK to remove this, however in future when we are publishing pre-built image to Dockerhub we may need to add it back. Also i dont see any problem in having this label as it is just a metadata to an image.

Comment (rajeevarakkal) [reply docker/Dockerfile:2]: @saishivarcr , Some point when we have the docker image posted, we may need to bring this back and it make sense that point of time. Let get it done that point of time.  Currently customer may opt to customize the script and chances are there customer will be adding multiple packages in to this script and the generate the docker build out of it. IMHO it is risky to take maintainer rights of any such images. If you still feels this needs to be kept, i would request you to get a sign off from Security Team. I am good in that case

Comment (saishivarcr) [reply docker/Dockerfile:2]: @rajeevarakkal I will remove this.

File: docker/README.md
Comment (saishivarcr) [reply docker/README.md:9]: We are shipping any docker image. Consumers of this file need to be aware of inherent security risks associated with docker images and containers. Will add relevant link to readme.

File: docker/README.md
Comment (saishivarcr) [reply docker/README.md:16]: will add link to example playbooks

File: docker/README.md
Comment (saishivarcr) [reply docker/README.md:26]: will update readme with mount instructions

File: docker/README.md
Comment (saishivarcr) [reply docker/README.md:6]: Base image link is added. Details on OMSDK and ansible packages are already covered in repository readme, hence its redundant to cover it over here.

Comment (saishivarcr) [reply docker/README.md:6]: will add the size of image.

**Comment by saishivarcr** [comment]
@rajeevarakkal I have incorporated the changes based on your feedback. Kindly review it.

File: docker/README.md
Comment (rajeevarakkal) [reply docker/README.md]: @saishivarcr; Better we should have the over all disclaimer stating that the respective open source component used in your script, before building the image out of it, get the required approvals from your management. Dell as such won't be responsible for in case anything. This is because the script is pulling out various open source bundles such as docker base image of python:slim3 and ansible which are not dellemc deliverables.

File: docker/README.md
Comment (rajeevarakkal) [reply docker/README.md:47]: @saishivarcr, yup.. lets be specific on what details we need/using from `Docker Host` since host is something shared by many dockers. Also the specific access rights we need on host. It should end up like customer should run generated images in privileged mode

Comment (saishivarcr) [reply docker/README.md:47]: @rajeevarakkal privileged mode is not required as share will not be mounted inside the container. Share has to be mounted on docker engine and bind mounted in the container at `/dellemc`. These instructions are already updated. Also default access rights are sufficient in this case, hence not documented.

File: docker/README.md
Comment (rajeevarakkal) [inline docker/README.md:21]: Any reasons why we are point out security vulnerabilities of docker images. ? What is relevance with the context? I think this line doesn't really useful or encouraging customer to use this.

**Review by rajeevarakkal — CHANGES_REQUESTED** [review]

File: docker/README.md
Comment (rajeevarakkal) [inline docker/README.md:20]: can we talk about 1 liner about Redhat ansible bundle as well?

**Review by rajeevarakkal — CHANGES_REQUESTED** [review]

File: docker/README.md
Comment (rajeevarakkal) [inline docker/README.md:20]: does this needed? `Change this value if you want to use a different python image as a base`.

File: docker/README.md
Comment (rajeevarakkal) [inline docker/README.md:33]: Can we mention where is available. Exactly what we asked in the below comments? Details of the Docker Host resources used.

**Review by rajeevarakkal — CHANGES_REQUESTED** [review]

File: docker/README.md
Comment (rajeevarakkal) [inline docker/README.md:49]: `playbook.yml` is this available by default?

**Review by rajeevarakkal — CHANGES_REQUESTED** [review]

File: docker/README.md
Comment (saishivarcr) [reply docker/README.md]: readme updated with this information

File: docker/README.md
Comment (saishivarcr) [reply docker/README.md]: removed this from readme

File: docker/README.md
Comment (saishivarcr) [reply docker/README.md]: details specific to Redhat are not required, in case of build failure appropriate action which the customer has to take is mentioned in the readme.

Comment (saishivarcr) [reply docker/README.md]: removed this from the readme.

File: docker/README.md
Comment (saishivarcr) [reply docker/README.md:46]: The customer has to either create the playbook or he can download the playbooks from samples. These files(playbook, inventory, and optional group var files) need to be available on the host machine which will be bind-mounted in the container.

File: docker/README.md
Comment (saishivarcr) [reply docker/README.md:34]: Readme update with the link to sample playbooks.

**Review by rajeevarakkal — APPROVED** [review]

**Review by Sajna-Shetty — APPROVED** [review]
looks good to approve

**Pull request merged**
